iml-agent/src/daemon_plugins/ostpool.rs

```diff
@@ -0,0 +1,225 @@
+
+#[derive(Debug)]
+pub struct PoolState {
+    state: Arc<Mutex<HashMap<String, HashMap<String, Vec<String>>>>>,
+}
+
+pub fn create() -> impl DaemonPlugin {
+    PoolState {
+        state: Arc::new(Mutex::new(HashMap::new())),
+    }
+}
+
+async fn list_fs() -> Result<Vec<String>, ImlAgentError> {
+    Ok(lctl(vec!["get_param", "-N", "mdt.*-MDT0000"]).await
+       .map(|o| {
+           String::from_utf8_lossy(&o.stdout)
+               .lines()
+               .filter_map(|line| {
+                   line.split('.')
+                       .skip(1)
+                       .next()
+                       .and_then(|mdt| mdt.split("-MDT").next() )
+               })
+               .map(|s| s.to_string())
+               .collect()
+       })
+       .unwrap_or(vec![]))
+}
+
+/// Return vector of changes between existing fs state and list of pools
+/// from current filesystem
+fn diff_state(
+    fs: String,
+    state: &mut HashMap<String, Vec<String>>,
+    pools: &Vec<OstPool>,
+) -> Result<Vec<(OstPoolAction, OstPool)>, ImlAgentError> {
+    let mut rmlist = state.clone();
+    let mut changes: Vec<(OstPoolAction, OstPool)> = vec![];
+    for pool in pools {
+        tracing::debug!("Updating {}.{}", fs, pool.name);
+
+        if let Some(oldosts) = rmlist.remove(&pool.name) {
+            let new = HashSet::from_iter(pool.osts.clone());
+            let old = HashSet::from_iter(oldosts);
+
+            let addlist: HashSet<_> = new.difference(&old).cloned().collect();
+            if !addlist.is_empty() {
+                state.insert(pool.name.clone(), old.union(&addlist).cloned().collect());
+                changes.push((
+                    OstPoolAction::Grow,
+                    OstPool {
+                        name: pool.name.clone(),
+                        filesystem: fs.clone(),
+                        osts: addlist.into_iter().collect(),
+                    },
+                ));
+            }
+
+            let remlist: Vec<String> = old.difference(&new).cloned().collect();
+            if !remlist.is_empty() {
+                changes.push((
+                    OstPoolAction::Shrink,
+                    OstPool {
+                        name: pool.name.clone(),
+                        filesystem: fs.clone(),
+                        osts: remlist,
+                    },
+                ));
+            }
+            state.insert(pool.name.clone(), pool.osts.clone());
+        } else {
+            state.insert(pool.name.clone(), pool.osts.clone());
+            changes.push((OstPoolAction::Add, pool.clone()));
+        }
+    }
+
+    // Removed pools
+    for name in rmlist.keys() {
+        let osts = state.remove(name.as_str()).unwrap_or(vec![]);
+        changes.push((
+            OstPoolAction::Remove,
+            OstPool {
+                name: name.clone(),
+                filesystem: fs.clone(),
+                osts,
+            },
+        ));
+    }
+    Ok(changes)
+}
+
+async fn init_fsmap(
```

Comment:
This takes a lock once for each call, and then another lock for each iteration.

Instead of updating the state each time, we can build up the data we need to populate the state and then take a single lock to update.

If we change this fn so it returns data instead of updating state:

```rust
async fn get_pool_info(fsname: String) -> Vec<(String, String, Vec<String>)> {
    let xs = pool_list(&fsname)
        .await
        .unwrap_or(vec![])
        .into_iter()
        .map(move |pool| {
            let fsname = fsname.clone();

            async {
                let osts = ost_list(&fsname, &pool).await.unwrap_or(vec![]);

                (fsname, pool, osts)
            }
        });

    join_all(xs).await
}
```

and add another fn that inits the state:

```rust
fn init_pools(
    state: &mut HashMap<String, HashMap<String, Vec<String>>>,
    xs: Vec<(String, String, Vec<String>)>,
) -> Vec<(OstPoolAction, OstPool)> {
    xs.into_iter()
        .inspect(|(fsname, pool, osts)| {
            state
                .entry(fsname.clone())
                .and_modify(|hm| {
                    hm.insert(pool.clone(), osts.clone());
                })
                .or_insert_with(|| HashMap::from_iter(vec![(pool.clone(), osts.clone())]));
        })
        .map(|(filesystem, name, osts)| {
            (
                OstPoolAction::Initial,
                OstPool {
                    name,
                    filesystem,
                    osts,
                },
            )
        })
        .collect()
}
```

we can use them to init / update state in the daemon.